### DISCUSSION TITLE
Auto-update doesn't work using Quadlet

### DISCUSSION THREAD

**Discussion by q7nm** [containers/podman#20675] [General]
### Issue Description

I can't use the auto-update feature for this container:
```
[Unit]
Description=Wireguard-ui container

[Container]
ContainerName=wireguard-ui
Image=docker.io/ngoduykhanh/wireguard-ui:latest
AutoUpdate=registry
AddCapability=NET_ADMIN
Environment=BIND_ADDRESS=localhost:80
PodmanArgs=--pod wireguard-pod

[Service]
Restart=always

[Install]
WantedBy=default.target
```

```
	* auto-updating container "f52be832eb2da504deb7f6ed482adb0247cf4f0a1d8a8db69ccee22112201d69": no PODMAN_SYSTEMD_UNIT label found
```

But everything is OK with this:
```
[Unit]
Description=SSH-cloudflared container

[Container]
ContainerName=ssh-cloudflared
Image=docker.io/cloudflare/cloudflared:latest
AutoUpdate=registry
Secret=ssh-tunnel,type=env,target=TUNNEL_TOKEN
Network=ssh-cloudflared
Exec=tunnel run

[Service]
Restart=always

[Install]
WantedBy=default.target
```

```
            ssh-cloudflared.service  7b357ee8e487 (ssh-cloudflared)  docker.io/cloudflare/cloudflared:latest  registry    false
```

I also have the same problem with other containers but I think it's not necessary to show more.

### Steps to reproduce the issue

Steps to reproduce the issue
1. Create the .container file.
2. Start the container.
3. Enter `podman auto-update`. 

### Describe the results you received

 I received non-working auto-update.

### Describe the results you expected

I would like to have the working auto-update feature.

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.32.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.8-2.fc39.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.8, commit: '
  cpuUtilization:
    idlePercent: 95.51
    systemPercent: 2.14
    userPercent: 2.35
  cpus: 2
  databaseBackend: boltdb
  distribution:
    distribution: fedora
    variant: cloud
    version: "39"
  eventLogger: journald
  freeLocks: 2033
  hostname: private-server
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 6.5.11-300.fc39.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 187219968
  memTotal: 992841728
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.8.0-1.fc39.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.8.0
    package: netavark-1.8.0-2.fc39.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.8.0
  ociRuntime:
    name: crun
    package: crun-1.11.1-1.fc39.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.11.1
      commit: 1084f9527c143699b593b44c23555fb3cc4ff2f3
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20231004.gf851084-1.fc39.x86_64
    version: |
      pasta 0^20231004.gf851084-1.fc39.x86_64
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: false
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.2-1.fc39.x86_64
    version: |-
      slirp4netns version 1.2.2
      commit: 0ee2d87523e906518d34a6b423271e4826f71faf
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 991686656
  swapTotal: 991948800
  uptime: 0h 13m 31.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /usr/share/containers/storage.conf
  containerStore:
    number: 7
    paused: 0
    running: 7
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 26208088064
  graphRootUsed: 1402945536
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Supports shifting: "true"
    Supports volatile: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 4
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.7.2
  Built: 1698762721
  BuiltTime: Tue Oct 31 14:32:01 2023
  GitCommit: ""
  GoVersion: go1.21.1
  Os: linux
  OsArch: linux/amd64
  Version: 4.7.2
```

### Podman in a container

No

### Privileged Or Rootless

Privileged

### Upstream Latest Release

Yes

### Additional information

Maybe the problem is the using of pods? I don't know if it's necessary to create them another way.
```
sudo podman pod create --name wireguard-pod --network host -v wireguard-db:/app/db -v /etc/wireguard:/etc/wireguard:z
```

**Comment by ygalblum** [comment]
Yes, as you've stated, this is related to the container being part of a Pod. According to this: https://github.com/containers/podman/blob/main/pkg/autoupdate/autoupdate.go#L453 once the container is part of a Pod, podman looks for the label at the Pod level and not at the container level.

So, this is a Podman general behavior and not a Quadlet one. @rhatdan @vrothberg any thoughts?

**Comment by rhatdan** [comment]
I think it is reasonable for users to expect the attribute to work if set on a container within a Pod, but want @vrothberg to chime in.

**Comment by vrothberg** [comment]
> I think it is reasonable for users to expect the attribute to work if set on a container within a Pod, but want @vrothberg to chime in.

Pods in Quadlet were meant to be used via `.kube` files.  The `PodmanArgs=--pod=xxx` works but in my eyes that really a hack.  It also implies that auto-updates with a Quadlet `.container` using this hack/workaround is not tested.

I personally prefer to not support this scenario at all and improve documentation of Quadlet/Podman.

This specific issue seems fairly easy to fix by changing the semantics of https://github.com/containers/podman/blob/main/pkg/autoupdate/autoupdate.go#L453 to return the container's unit (again) if the infra container is not part of another unit.  But again, there is a reason why Quadlet does not support a `Pod` field and supporting auto-updates in this scenario would mean that `.container`+pod is somehow supported.  So, although fixing the issue is probably easy, it may be better to improve the docs.

**Reply by haeckle** [reply]
As pods will be supported by Quadlet in the [next version](https://github.com/containers/podman/pull/20762#issuecomment-1880969803) are there now any plans to bring support for auto updating to those pods or individual containers?

**Reply by rhatdan** [reply]
@vrothberg WDYT?

**Reply by vrothberg** [reply]
I didn't test it but it should just work.  `.container` file can already make use of auto-update and can also be linked to a `.pod`.

**Reply by haeckle** [reply]
Thanks! Will try when it lands in fedora stable.

**Comment by q7nm** [comment]
> Pods in Quadlet were meant to be used via `.kube` files

Is it rational to use K8s for my little tasks? I've never used it, so I don't know.

**Comment by rhatdan** [comment]
If you run your pod/container via the podman command and then do a `podman kube generate`, theoretically you should have a kube.yaml file ready to run within a quadlet.

**Reply by q7nm** [reply]
That's interesting, thx.

**Comment by rhatdan** [comment]
Since we already described a workaround and our desire not to support, moved to a discussion.

**Reply by q7nm** [reply]
I have a question. Does it generate a kube.yaml w/o my secrets? It seems I have to manually add them.

**Comment by 459below** [comment]
I solved the problem of: "_I want my containers inside my pod to update their image when calling_ `podman auto-update`" by adding the label to the K8s deployment.yml like this:

```
spec:
(...)
  template:
     (...)
      labels:
        io.kompose.network/ejabberd-default: "true"
        io.kompose.service: ejabberd
        **io.containers.autoupdate: registry**
    spec:
      containers:
        - image: ghcr.io/processone/ejabberd:23.04
(...)
```

Now it triggers an update like this
```
ejabberd$ podman auto-update 
            UNIT              CONTAINER                             IMAGE                              POLICY      UPDATED
            ejabberd.service  b7bd22ba32a3 (ejabberd-pod-ejabberd)  ghcr.io/processone/ejabberd:23.04  registry    false
```

Maybe this has been stated already in the thread but it was not entirely clear to me, so I wanted to state it here.

Thanks Ygal Blum. What you said really helped me here:
_once the container is part of a Pod, podman looks for the label at the Pod level and not at the container level._

**Comment by chris42** [comment]
As I am just trying to get auto-update running on a 4.9.3 version, I have two short questions: Currently I have the label and the env set on container level, still get the error "no PODMAN_SYSTEMD_UNIT label found". 
As I understood this thread here, I should set it on pod level. Two things:

1. It does not work. I still get the PODMAN_SYSTEMD_UNIT error. The error also points to setting the env for the service name, however that is not possible on pod level?
2. How is distinguished between a container with registry image and local image in the same pod? Do I need to set the label additionally on container level as well?

Is there actually a working  setup for this in 4.9.3? How does it have to look?
